text/0000-fmt-size-hint.md

```diff
@@ -0,0 +1,106 @@
+- Start Date: 2015-01-13
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Add a `size_hint` method to each of the `fmt` traits, allowing a buffer to allocate with the correct size before writing.
+
+# Motivation
+
+Using the `fmt` traits is slower than a straight `memcpy` of data. The removal of `std::is_utf8` has helped some. The other low-hanging fruit is to add a size hint, so as to reduce or prevent unnecessary allocations during writing to the output buffer. My initial implementation includes [benchmarks][], which looked like this on machine:
+
+[benchmarks]: https://gist.github.com/seanmonstar/8fb7aa6b0512b80522f9#file-size_hint-rs-L91-L162
+
+```
+running 11 tests
+test bench_long         ... bench:       133 ns/iter (+/- 18)
+test bench_long_hint    ... bench:        72 ns/iter (+/- 10)
+test bench_long_memcpy  ... bench:        44 ns/iter (+/- 2)
+test bench_med          ... bench:       112 ns/iter (+/- 10)
+test bench_med_hint     ... bench:        59 ns/iter (+/- 7)
+test bench_med_memcpy   ... bench:        32 ns/iter (+/- 6)
+test bench_nested       ... bench:       248 ns/iter (+/- 19)
+test bench_nested_hint  ... bench:       134 ns/iter (+/- 6)
+test bench_short        ... bench:        96 ns/iter (+/- 13)
+test bench_short_hint   ... bench:        60 ns/iter (+/- 3)
+test bench_short_memcpy ... bench:        33 ns/iter (+/- 3)
+```
+
+# Detailed design
+
+Add a `size_hint` method to all of the `fmt` traits, with a default implementation so no one is broken. Opting in simply means better performance. All traits should have their own size_hint implementation, since the trait used can change the length of the output written.
+
+```rust
+trait Show {
+    fn fmt(&self, &mut Formatter) -> Result;
+    fn size_hint(&self) -> SizeHint {
+        SizeHint { min: 0, max: None }
+    }
+}
+```
+
+Add a `SizeHint` type, with named properties, instead of using tuple indexing. Include an `Add` implementation for `SizeHint`, so they can be easily added together from nested properties.
```

Comment:
This is an interesting deviation from the size hints of iterators. I would personally expect the two to have the same signature, and this may want to at least discuss the discrepancy between the two return values.